source/azure_iot_provisioning_client.c

```diff
@@ -0,0 +1,963 @@
+            return;
+        }
+
+        if( ( xReturn =
+                  prvAzureIoTProvisioningClientRequestPayload( xAzureIoTProvisioningClientHandle,
+                                                               xAzureIoTProvisioningClientHandle->_internal.pucProvisioningScratchBuffer +
+                                                               xMQTTTopicLength,
+                                                               xAzureIoTProvisioningClientHandle->_internal.ulProvisioningScratchBufferLength -
+                                                               ( uint32_t ) xMQTTTopicLength,
+                                                               &xMQTTPayloadLength ) ) != eAzureIoTProvisioningSuccess )
+        {
+            prvAzureIoTProvisioningClientUpdateState( xAzureIoTProvisioningClientHandle, xReturn );
+            return;
+        }
+
+        ( void ) memset( ( void * ) &xMQTTPublishInfo, 0x00, sizeof( xMQTTPublishInfo ) );
+        xMQTTPublishInfo.xQOS = eAzureIoTMQTTQoS0;
+        xMQTTPublishInfo.pcTopicName = xAzureIoTProvisioningClientHandle->_internal.pucProvisioningScratchBuffer;
+        xMQTTPublishInfo.usTopicNameLength = ( uint16_t ) xMQTTTopicLength;
+        xMQTTPublishInfo.pvPayload = xAzureIoTProvisioningClientHandle->_internal.pucProvisioningScratchBuffer + xMQTTTopicLength;
+        xMQTTPublishInfo.xPayloadLength = xMQTTPayloadLength;
+        usPublishPacketIdentifier = AzureIoTMQTT_GetPacketId( &( xAzureIoTProvisioningClientHandle->_internal.xMQTTContext ) );
+
+        if( ( xResult = AzureIoTMQTT_Publish( &( xAzureIoTProvisioningClientHandle->_internal.xMQTTContext ),
+                                              &xMQTTPublishInfo, usPublishPacketIdentifier ) ) != eAzureIoTMQTTSuccess )
+        {
+            AZLogError( ( "Failed to publish. Error=%d", xResult ) );
+            xReturn = eAzureIoTProvisioningPublishFailed;
+        }
+        else
+        {
+            xReturn = eAzureIoTProvisioningSuccess;
+        }
+
+        prvAzureIoTProvisioningClientUpdateState( xAzureIoTProvisioningClientHandle, xReturn );
+    }
+    else
+    {
+        AZLogWarn( ( "Request action called in wrong state %d\r\n",
+                     xAzureIoTProvisioningClientHandle->_internal.ulWorkflowState ) );
+    }
+}
+/*-----------------------------------------------------------*/
+
+/**
+ *
+ * Implementation of parsing response action, this action is only allowed in azureiotprovisioningWF_STATE_RESPONSE
+ *
+ * */
+static void prvAzureIoTProvisioningClientParseResponse( AzureIoTProvisioningClientHandle_t xAzureIoTProvisioningClientHandle )
+{
+    az_result xCoreResult;
+    az_span xPayload = az_span_create( xAzureIoTProvisioningClientHandle->_internal.pucProvisioningLastResponsePayload,
+                                       ( int32_t ) xAzureIoTProvisioningClientHandle->_internal.xProvisioningLastResponsePayloadLength );
```

Comment:
Another nit around length - sorry - but again `xProvisioningLastResponsePayloadLength ` does that need to be so long?  The fact it's internal to a handle about provisioning makes me think this borders on repeating ourselves?